### PR TITLE
[jobs] Make local pip/conda requirements files work with jobs

### LINKED ISSUE #22044: Support `runtime_env` `pip` `requirements.txt` in Job submission
Same for conda.  Most likely the reason this doesn't work already is that the runtime env gets set up on the remote cluster, but the `requirements.txt` or `environment.yml` file is not present on the cluster.

Option 1: Push the files to the cluster before the runtime_env gets built
Option 2: Parse the file to List or JSON before sending the runtime_env to the cluster

Option 2 should be more straightforward.  Actually, this is how we handle this situation already outside the Job submission context.

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?
Previously, when using a runtime env with job submission, conda and pip requirements file paths were passed to the cluster unmodified.  When the remote job server tried to install the runtime environment, the installation would fail because the file only existed on the local machine, not the remote cluster.

The approach taken in this PR is to parse and validate the runtime env on the client side before sending it to the job server over HTTP.  By doing this, the requirements files are parsed into a list (for pip) or dict (for conda).

One consequence is that validation errors now raise exceptions and cause the job to not be submitted.  This is a change in the semantics of job submission so it's not clear if the pros outweigh the cons.  One con is that job managers will have to handle exceptions instead of just checking the job status.

## Related issue number

Closes #22044
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: dashboard/modules/job/job_manager.py [hunk 1]
```diff
@@ -137,12 +138,21 @@ def _exec_entrypoint(self, logs_path: str) -> subprocess.Popen:
                 terminated or killed upon user calling stop().
         """
         with open(logs_path, "w") as logs_file:
+            if self._runtime_env.get("pip") is not None:
+                # Replace "python" in the user's entrypoint script with the
+                # Python executable for the present runtime env.
+                cmd = f'python() {{ "{sys.executable}" "$@" ; }} && ' + self._entrypoint
+                executable = "/bin/bash"
```
Comment (edoakes) [inline dashboard/modules/job/job_manager.py:145]: hmm why is this necessary?

File: dashboard/modules/job/sdk.py [hunk 2]
```diff
@@ -296,6 +297,8 @@ def submit_job(
         metadata.update(self._default_metadata)
 
         self._upload_working_dir_if_needed(runtime_env)
+        # Run the RuntimeEnv constructor to parse local pip/conda requirements files.
+        runtime_env = RuntimeEnv(**runtime_env).to_dict()
```
Comment (edoakes) [inline dashboard/modules/job/sdk.py:301]: Do this constructor also do working_dir and py_modules uploading? For example, if this and the previous line were reversed what would happen?

Ideally we would unify the working_dir uploading part so we don't have these divergent codepaths. For example, we could pass in an optional `upload_working_dir` callback into the constructor, which in this case would use HTTP.

**Review by edoakes — COMMENTED** [review]
A little concerned about the entrypoint modification part -- can you explain why that's necessary / what happens if we leave it out?

File: dashboard/modules/job/job_manager.py (hunk 1 above)
Comment (architkulkarni) [reply dashboard/modules/job/job_manager.py:145]: Otherwise an entrypoint script that uses `python` won't be able to import the pip packages, so the test `"python -c 'import pip_install_test'"` will fail.  (The driver scripts will also fail their imports outside of tasks and actors, it's the same as the Ray client server issue we discussed.)

The reason is that `pip` runtime_env actually changes the `python` executable that it uses for tasks and actors.

File: dashboard/modules/job/sdk.py (hunk 2 above)
Comment (architkulkarni) [reply dashboard/modules/job/sdk.py:301]: The constructor doesn't do any uploading, it always assumes everything is a URI already in the input.  So if you switched the lines, it would fail.  I think this status quo is okay, but I agree we ideally wouldn't have to always call upload_py_modules and upload_working_dir before calling the runtime env constructor.  Maybe we can refactor in the future

**Comment by architkulkarni** [comment]
@edoakes Any other concerns about this?  I agree the bash is a little weird, but I couldn't think of a better solution.  I think the workflow it fixes is a basic requirement (pass in `pip` runtime env, and be able to import it in a driver or in a non-Ray script).  However there are still some gaps; for example if the user entrypoint executes a second script which calls `python` inside it, then this `python` won't be replaced by the correct executable.

File: dashboard/modules/job/job_manager.py (hunk 1 above)
Comment (architkulkarni) [reply dashboard/modules/job/job_manager.py:145]: Discussed offline, this is a hack that won't work in common scenarios like `rllib run`.  The underlying issue is specific to runtime envs, tracked here https://github.com/ray-project/ray/issues/22968.

Because that issue is independent of the one in this PR (fixing requirements files), I'll remove this hack from this PR so that just the part about `requirements.txt` can be merged independently.  After https://github.com/ray-project/ray/issues/22968 is fixed we can add back the `python -c "import pip-install-test"` unit test.

**Comment by architkulkarni** [comment]
The bash workaround has been reverted and the test has been replaced by one which only imports inside a Ray task.  Tests passed earlier so they should pass again and this PR should be ready to merge.  

After https://github.com/ray-project/ray/pull/22983 is merged I will add back the jobs test with entrypoint `python -c "import pip_install_test"`.

**Comment by architkulkarni** [comment]
tests:test_actor_client_mode,  Windows serve:test_ray_client, Windows tests:test_object_spilling broken on master, mac tests stalled on master

**Comment by edoakes** [comment]
@architkulkarni merge conflicts from other PR getting merged

**Review by edoakes — APPROVED** [review]

**Comment by architkulkarni** [comment]
Windows test_serve flaky on master
